x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java

```diff
@@ -439,20 +439,20 @@ private void buildUser(SamlAttributes attributes, ActionListener<AuthenticationR
             return;
         }
 
-        final Map<String, Object> userMeta = new HashMap<>();
+        final Map<String, Object> userMetaBuilder = new HashMap<>();
         if (populateUserMetadata) {
             for (SamlAttributes.SamlAttribute a : attributes.attributes()) {
-                userMeta.put("saml(" + a.name + ")", a.values);
+                userMetaBuilder.put("saml(" + a.name + ")", a.values);
                 if (Strings.hasText(a.friendlyName)) {
-                    userMeta.put("saml_" + a.friendlyName, a.values);
+                    userMetaBuilder.put("saml_" + a.friendlyName, a.values);
                 }
             }
         }
         if (attributes.name() != null) {
-            userMeta.put(USER_METADATA_NAMEID_VALUE, attributes.name().value);
-            userMeta.put(USER_METADATA_NAMEID_FORMAT, attributes.name().format);
+            userMetaBuilder.put(USER_METADATA_NAMEID_VALUE, attributes.name().value);
+            userMetaBuilder.put(USER_METADATA_NAMEID_FORMAT, attributes.name().format);
         }
-
+        final Map<String, Object> userMeta = Map.copyOf(userMetaBuilder);
```

Comment:
Okay, looking at how `userMetaBuilder` is populated I thought that the stream API (that builds the unmodifiable map) is unsuitable.
I'll give it a try though...

Reply:
I originally did not understood the suggestion, sorry Yogesh!
Yes, I prefer to create the `AbstractImmutableCollection` as soon as possible to avoid the unnecessary wrappings in `Collections.unmodifiableMap` and duplicated `null` filterings later on. Plus, the goal is to pass the `userMeta` to the `UserRoleMapper.UserData` not the `User`.